frontend/ui/widget/qrwidget.lua

```diff
@@ -0,0 +1,38 @@
+--[[
+QRWidget shows a QR code for a given text.
+]]
+
+local ImageWidget = require("ui/widget/imagewidget")
+local RenderImage = require("ui/renderimage")
+local Widget = require("ui/widget/widget")
+local logger = require("logger")
+local qrencode = require("qrencode")
+
+local QRWidget = Widget:new{
+    scale_factor = 0,
+    text = ""
+    -- see ImageWidget for other options.
+}
+
+function QRWidget:init()
+    local ok, ret = qrencode.qrcode(self.text)
+    if not ok then
+        logger.info("QRWidget: failed to generate QR code.")
+        return
+    else
+        -- We generate pbm data
+        local qr = {"P1\n", #ret[1], " ", #ret, "\n"}
+        for _, col in ipairs(ret) do
+            for _, lgn in ipairs(col) do
+                table.insert(qr, lgn > 0 and "1 " or "0 ")
+            end
+            table.insert(qr, "\n")
+        end
+        qr = table.concat(qr, '')
+        local dim = math.min(self.width, self.height)
+        self.image = RenderImage:renderImageData(qr, #qr, dim, dim)
+        return ImageWidget:new(self)
```

Comment:
I'm not sure this is a proper way to use our widgets :new()/:init(). `QRWidget:init()` does not init a proper QRWidget, but returns an ImageWidget. (May be it works, but it's a bit confusing vs all our other widgets I think).
If you dont plan on using a QRWidget anyplace else, you might as well just add it as a local function in qrmessage.lua `local getQRImageWidget = function(...)`
Otherwise, QRWidget might probably _extend_ ImageWidget to be more proper.

Reply:
+1 on the extend proposal ;).